csrc/preseg_passes/mark_aliases_prepare.cpp

```diff
@@ -50,86 +110,100 @@ void MarkAliasesPreparePass::runPass(Fusion* fusion) {
+  //
+  //   N/M -> M1 -> N/M
+  //          |
+  //          -> M2
+  //
+  // we want to avoid putting a `segment_set` before M1, a meta op, because
+  // that would lead to two kernels. See AliasTest.DoNotOverSegment_* for more
+  // examples.
+  const std::unordered_set<Expr*>& used_by_non_aliases =
+      exprsUsedByNonAliases(analysis, fusion);
+  std::vector<Use> uses_to_segment;
+  uses_to_segment.reserve(fusion->outputs().size());
+  for (auto* out : ir_utils::filterByType<TensorView>(fusion->outputs())) {
+    Use use_to_segment = findUseToSegment(out, analysis, used_by_non_aliases);
+    if (use_to_segment.use_of != out) {
+      uses_to_segment.push_back(use_to_segment);
     }
   }
 
-  // Mark all expressions that are (transitively) used by non-alias outputs.
-  auto used_by_non_aliases =
-      [](const std::vector<Expr*>& exprs) -> std::unordered_set<Expr*> {
-    return {exprs.begin(), exprs.end()};
-  }(StmtSort::getExprsTo(non_alias_outs));
-
-  for (TensorView* aliased_io : aliased_ios) {
-    // Divide the users of aliased_io into two groups according to
-    // `used_by_non_aliases`.
-    std::vector<Expr*> users_used_by_non_aliases;
-    std::vector<Expr*> users_used_only_by_aliases;
-    for (Expr* e : aliased_io->uses()) {
-      if (used_by_non_aliases.count(e)) {
-        users_used_by_non_aliases.push_back(e);
-      } else {
-        users_used_only_by_aliases.push_back(e);
-      }
+  // The remaining are optimizations to reduce the number of `segment_set`s
+  // inserted.
+  //
+  // Group `uses_to_segment` by `use_of` and remove duplicates.
+  std::sort(uses_to_segment.begin(), uses_to_segment.end());
+  uses_to_segment.erase(
+      std::unique(uses_to_segment.begin(), uses_to_segment.end()),
+      uses_to_segment.end());
+
+  if (isDebugDumpEnabled(DebugDumpOption::PreSegmenterLogging)) {
+    for (const auto& use : uses_to_segment) {
+      debug() << "Will put a segment_set at " << use.user << std::endl;
     }
+  }
 
-    // If all users are used by non-alias outputs, do nothing. Adding
-    // segment_set around unlikely creates more no-op regions.
-    if (users_used_only_by_aliases.empty()) {
-      continue;
+  auto i = uses_to_segment.begin();
+  while (i != uses_to_segment.end()) {
+    TensorView* use_of = i->use_of;
+    auto j = i;
+    while (j != uses_to_segment.end() && j->use_of == use_of) {
+      j++;
     }
 
-    // If all users are used by aliases, put a `segment_set` before it.
-    if (users_used_by_non_aliases.empty()) {
-      if (aliased_io->isFusionInput()) {
-        // A `segment_set` before a fusion input is useless.
-        continue;
-      }
-      // Rarely, if `aliased_io` is already defined by `segment_set`, don't
-      // create another `segment_set`.
-      if (LoadStoreOp* def =
-              dynamic_cast<LoadStoreOp*>(aliased_io->definition())) {
-        if (def->opType() == LoadStoreOpType::SegmenterSet) {
-          continue;
+    auto insert_segment_set = [&]() {
```

Comment:
dumb question, this lambda is not used anywhere else. Does lambda bring anything to the table?

Reply:
So I can early-return inside the lambda without exiting MarkAliasesPreparePass::runPass, the outer function. Alternatively, I could use nested ifs or say
```
if (...) {
  i = j;
  continue;
}
```
which I'm not a big fan of.

Reply:
FYI, I'll make insert_segment_set a real function. The capture list is much shorter than I expected.